src/adservice/src/main/java/hipstershop/AdService.java

```diff
@@ -0,0 +1,208 @@
+        if (req.getContextKeysCount() > 0) {
+          span.addAnnotation(
+              "Constructing Ads using context",
+              ImmutableMap.of(
+                  "Context Keys",
+                  AttributeValue.stringAttributeValue(req.getContextKeysList().toString()),
+                  "Context Keys length",
+                  AttributeValue.longAttributeValue(req.getContextKeysCount())));
+          for (int i = 0; i < req.getContextKeysCount(); i++) {
+            Ad ad = service.getAdsByKey(req.getContextKeys(i));
+            if (ad != null) {
+              ads.add(ad);
+            }
+          }
+        } else {
+          span.addAnnotation("No Context provided. Constructing random Ads.");
+          ads = service.getDefaultAds();
+        }
+        if (ads.isEmpty()) {
+          // Serve default ads.
+          span.addAnnotation("No Ads found based on context. Constructing random Ads.");
+          ads = service.getDefaultAds();
+        }
+        AdResponse reply = AdResponse.newBuilder().addAllAds(ads).build();
+        responseObserver.onNext(reply);
+        responseObserver.onCompleted();
+      } catch (StatusRuntimeException e) {
+        logger.log(Level.WARNING, "GetAds Failed", e.getStatus());
+        return;
+      }
+    }
+  }
+
+  static final HashMap<String, Ad> cacheMap = new HashMap<String, Ad>();
+
+  Ad getAdsByKey(String key) {
+    return cacheMap.get(key);
+  }
+
+
+  public List<Ad> getDefaultAds() {
+    List<Ad> ads = new ArrayList<>(MAX_ADS_TO_SERVE);
+    Object[] keys = cacheMap.keySet().toArray();
+    for (int i=0; i<MAX_ADS_TO_SERVE; i++) {
+      ads.add(cacheMap.get(keys[new Random().nextInt(keys.length)]));
+    }
+    return ads;
+  }
+
+  public static AdService getInstance() {
+    return service;
+  }
+
+  /** Await termination on the main thread since the grpc library uses daemon threads. */
+  private void blockUntilShutdown() throws InterruptedException {
+    if (server != null) {
+      server.awaitTermination();
+    }
+  }
+
+  static void initializeAds() {
+    String adsUrl = System.getenv("ADS_URL");
+    cacheMap.put("camera", Ad.newBuilder().setRedirectUrl(adsUrl + "/camera")
+        .setText("MyPro camera for sale. 50% off.").build());
+    cacheMap.put("bike", Ad.newBuilder().setRedirectUrl(adsUrl + "/bike")
+        .setText("ZoomZoom bike for sale. 10% off.").build());
+    cacheMap.put("kitchen", Ad.newBuilder().setRedirectUrl(adsUrl + "/kitchen")
+        .setText("CutPro knife for sale. Buy one, get second set for free").build());
+    logger.info("Default Ads initialized");
+  }
+
+  /** Main launches the server from the command line. */
+  public static void main(String[] args) throws IOException, InterruptedException {
+    // Add final keyword to pass checkStyle.
+
+    initializeAds();
+
+    // Registers all RPC views.
+    RpcViews.registerAllViews();
+
+    // Registers logging trace exporter.
+    LoggingTraceExporter.register();
+
+    try {
```

Comment:
network might be unavailable during the first few milliseconds of a process in Istio. therefore we retry on other languages (3 times, 10*n seconds) while trying to reach stackdriver upon process startup.

Reply:
Added 3 tries 10 seconds apart.